### PR TITLE
Min s-t cut

### PR DESCRIPTION
**Pull request by jkinable**
Completely reworked the Minimum s-t cut algorithm:
- it now works in combination with all flow algorithms (Edmonds Karp/Push Relabel) and in combination with both directed and undirected graphs.
- it is significantly faster
- the unit tests have been extended to deal with an extensive amount of hard to find bugs.

Something related which we may want to address in a separate pull request:
The names of some of the methods in MaximumFlowAlgorithm are a bit confusing and not consistent between classes.
Currently we have the following methods:

MaximumFlow<E> buildMaximumFlow(V source, V sink);
Map<E, Double> getMaximumFlow()
double getMaximumFlowValue()
V getFlowDirection(E e)
double calculateMaximumFlow(V source,V sink)

You would expect that  "getMaximumFlow" returns a "MaximumFlow" object, but it returns a Map<E, Double>. Instead you need to invoke "buildMaximumFlow".
Hence I would propose to rename buildMaximumFlow -> getMaximumFlowValue, and to rename getMaximumFlowValue -> getFlowMap or getEdgeFlowMap or getFlowPerEdgeMap or something along those lines.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/MinSourceSinkCut.java [hunk 1]
```diff
@@ -30,116 +30,143 @@
...
+                    processQueue.add(directedGraph.getEdgeSource(edge));
                 }
             }
-            for (E edge : inEdges) {
-                processQueue.add(Graphs.getOppositeVertex(graph, edge, vertex));
+        }
+    }
+
+    public void processUndirectedGraph(MaximumFlow<E> maxFlow){
+//        TODO check implementation flow algorithm for undirected graphs. Seems wrong
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/MinSourceSinkCut.java]: Time to remove this and the println below, right?

Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/MinSourceSinkCut.java]: Hm, this is the wrong version. Must have messed something up with version control. Will revert this file back to the original.

**Comment by jkinable** [comment]
What do you think about my suggestion to rename the max flow code:

You would expect that "getMaximumFlow" returns a "MaximumFlow" object, but it returns a Map. Instead you need to invoke "buildMaximumFlow".
Hence I would propose to rename buildMaximumFlow -> getMaximumFlow, and to rename getMaximumFlowValue -> getFlowMap or getEdgeFlowMap or getFlowPerEdgeMap or something along those lines.
This would improve consistency with the other algorithm implementations.

**Comment by jsichi** [comment]
Thanks, merging now.

Regarding naming:  in the last release, the only interface method present was buildMaximumFlow.  So as long as that stays around, we should be free to make any other changes as appropriate since the new methods haven't been released yet.  @alexeykudinkin is that correct?

**Comment by alexeykudinkin** [comment]
Not sure that i quite got your point @jsichi. Yeah, your statement is correct. Though `buildMaximumFlow` doesn't please me either, and it still stay there for a historical reasons.

**Pull request merged**